File: 6/can_place_flowers.py
class Solution(object):
    def canPlaceFlowers(self, flowerbed, n):
        """
        :type flowerbed: List[int]
        :type n: int
        :rtype: bool
        """
        if len(flowerbed) == 0:
            return False
        if len(flowerbed) == 1:
            return True if n == 1 and not flowerbed[0] or n == 0 else False

        c = 0
        for i in range(len(flowerbed)):
            if flowerbed[i] == 1:
                continue

            if i == 0 and flowerbed[i + 1] == 0:
                c += 1
                flowerbed[i] = 1
                continue
            if i == len(flowerbed) - 1 and flowerbed[i - 1] == 0:
                c += 1
                flowerbed[i] = 1
                continue
            if flowerbed[i - 1] == 0 and flowerbed[i + 1] == 0:
                c += 1
                flowerbed[i] = 1
                continue

        return n <= c


s = Solution()
print(s.canPlaceFlowers([1, 0, 0, 0, 0, 1], 2))

File: 6/can_place_flowers.go
package main

import (
	"fmt"
)

func canPlace(flowerbed []int, i int) bool {
	if flowerbed[i] == 1 {
		return false
	}

	if i-1 >= 0 && i+1 < len(flowerbed) {
		if flowerbed[i-1] == 0 && flowerbed[i+1] == 0 {
			flowerbed[i] = 1
			return true
		} else {
			return false
		}
	}

	if i-1 < 0 && i+1 >= len(flowerbed) {
		return true
	}

	if i-1 < 0 {
		if flowerbed[i+1] == 0 {
			flowerbed[i] = 1
			return true
		}
	}

	if i+1 >= len(flowerbed) {
		if flowerbed[i-1] == 0 {
			flowerbed[i] = 1
			return true
		}
	}
	return false
}

func canPlaceFlowers(flowerbed []int, n int) bool {
	c := 0
	for i := 0; i < len(flowerbed); i += 1 {
		if canPlace(flowerbed, i) {
			c += 1
		}
	}
	return n <= c
}

func main() {
	f := []int{0}
	fmt.Println(canPlaceFlowers(f, 1))
	f = []int{1}
	fmt.Println(canPlaceFlowers(f, 0))
}
